kore/src/Kore/Internal/MultiAnd.hs

```diff
@@ -176,30 +150,28 @@ included in the Ord instance, items containing @\\forall@ and
 @\\exists@ may be considered inequal although they are equivalent in
 a logical sense.
 -}
-filterUnique :: Ord a => MultiAnd a -> MultiAnd a
-filterUnique = MultiAnd . Set.toList . Set.fromList . getMultiAnd
 
-{- | 'filterGeneric' simplifies a MultiAnd according to a function which
-evaluates its children to true/false/unknown.
+-- | 'make' constructs a simplified/normalized 'MultiAnd'.
+make :: (Ord term, TopBottom term) => [term] -> MultiAnd term
+make = foldAndPatterns . Set.fromList
+
+{- | 'foldAndPatterns' simplifies a set of children according to the `patternToMaybeBool`
+function which evaluates to true/false/unknown.
 -}
-filterGeneric ::
-    (child -> AndBool) ->
-    MultiAnd child ->
+foldAndPatterns ::
+    (Ord child, TopBottom child) =>
+    Set child ->
     MultiAnd child
-filterGeneric andFilter (MultiAnd patts) =
-    go andFilter [] patts
+foldAndPatterns patts = Set.foldr go mempty patts
   where
-    go ::
-        (child -> AndBool) ->
-        [child] ->
-        [child] ->
-        MultiAnd child
-    go _ filtered [] = MultiAnd (reverse filtered)
-    go filterAnd' filtered (element : unfiltered) =
-        case filterAnd' element of
-            AndFalse -> MultiAnd [element]
-            AndTrue -> go filterAnd' filtered unfiltered
-            AndUnknown -> go filterAnd' (element : filtered) unfiltered
+    go element mand =
+        case patternToMaybeBool element of
+            Just False -> MultiAndBottom element
+            Just True -> mand
+            Nothing -> case mand of
+                MultiAnd es -> MultiAnd $ Set.insert element es
+                MultiAndTop -> MultiAnd $ Set.singleton element
+                bottom -> bottom
```

Comment:
Is it worth swapping the `case` expressions here? Would mean the first `MultiAndBottom` will short-cut the further element evaluation (at the price of code repetition, because we cannot )
```suggestion
    go element mand =
        case mand of
            MultiAndBottom _ -> mand
            MultiAndTop -> single
            MultiAnd _ -> mand <> single -- not sure about this
    ~single =
        maybe
            (MultiAnd $ Set.singleton element)
            (\b -> if b then MultiAndTop else MultiAndBottom element)
            (patternToMaybeBool element)
```

Reply:
Oh, I wanted to short-circuit on `element`, however I forgot again that everything is strict. Changing to `go element ~mand` made performance much closer to master, but still no real improvement :/ Do you think short circuiting on the MultiAnd instead would be better?

Reply:
btw here is the table:

```
 Test                                            | lazy time          | ac0e3c0c403bbca4d812d921f8783f4dfcdbcb29-ghc8107 time | (lazy/ac0e3c0c403bbca4d812d921f8783f4dfcdbcb29-ghc8107) time 
-------------------------------------------------+--------------------+-------------------------------------------------------+--------------------------------------------------------------
 erc20/hkg/approve-spec.k                        | 157.71             | 152.25                                                | 1.0358620689655174                                           
 specs/benchmarks/ecrecover00-sigvalid-spec.k    | 333.37             | 321.45                                                | 1.037081972312957                                            
 specs/benchmarks/requires01-a0gt0-spec.k        | 129.13             | 124.38                                                | 1.0381894195208232                                           
 specs/benchmarks/storagevar02-nooverflow-spec.k | 115.31             | 110.99                                                | 1.0389224254437337                                           
 erc20/hkg/balanceOf-spec.k                      | 96.91              | 93.27                                                 | 1.0390264822558164                                           
 specs/mcd/cat-exhaustiveness-spec.k             | 147.39             | 141.68                                                | 1.0403020892151325                                           
 specs/benchmarks/address00-spec.k               | 126.75             | 121.83                                                | 1.040384141836986                                            
 specs/mcd/dsvalue-read-pass-spec.k              | 267.7              | 257.13                                                | 1.0411076109361024                                           
 specs/benchmarks/encode-keccak00-spec.k         | 319.56             | 306.52                                                | 1.0425420853451652                                           
 specs/functional/lemmas-spec.k                  | 35.5               | 33.97                                                 | 1.0450397409478953                                           
 erc20/hkg/transfer-success-1-spec.k             | 343.83             | 328.57                                                | 1.0464436801899137                                           
 erc20/hkg/transfer-failure-1-spec.k             | 280.78             | 266.37                                                | 1.0540976836730862                                           
 specs/benchmarks/staticloop00-a0lt10-spec.k     | 221.33             | 209.04                                                | 1.0587925755836205                                           
 specs/benchmarks/staticarray00-spec.k           | 176.83             | 164.6                                                 | 1.0743013365735117                                           
 TOTAL                                           | 2752.0999999999995 | 2632.0499999999997                                    | 1.0456108356604168
 ```
 
 its obviously a lot shorter but still clearly shows this change is at best no better than master

Reply:
> Oh, I wanted to short-circuit on `element`, [..] Do you think short circuiting on the MultiAnd instead would be better?
Actually not sure any more... If we had used a `foldl'` it would be better but this is a `foldr`. With the lazy `~mand` it _should_  short-circuit on the element.